Improve sample registration dialog

KochTobi opened the pull request.

KochTobi commented:
The dialog is ready and only has to be provided with the matching information in the correct place.

Steffengreiner commented on user-interface/src/main/java/life/qbic/datamanager/views/projects/project/samples/SampleDetailsComponent.java:
```diff
@@ -96,14 +102,23 @@ public class SampleDetailsComponent extends PageArea implements Serializable {
+        sampleInformationService, this.batchRegistrationService,
+        this.sampleRegistrationService);
+
+    Button registerButton2 = new Button("Register", this::onRegisterButtonClicked);
```
There should only be one registerButton 
```suggestion
    Button registerButton = new Button("Register", this::onRegisterButtonClicked);
```

Steffengreiner commented on user-interface/src/main/java/life/qbic/datamanager/views/projects/project/samples/SampleDetailsComponent.java:
```diff
@@ -130,6 +145,57 @@ private void initButtonAndFieldBar() {
+        .flatMap(experimentInformationService::find)
+        .orElseThrow();
+
+    BatchRegistrationDialog2 dialog = new BatchRegistrationDialog2(
```
There should only be one BatchRegistrationDialog 💍 
```suggestion
    BatchRegistrationDialog dialog = new BatchRegistrationDialog2(
```

KochTobi replied:
There is now.

Steffengreiner commented on user-interface/src/main/java/life/qbic/datamanager/views/projects/project/samples/SampleDetailsComponent.java:
```diff
@@ -130,6 +145,57 @@ private void initButtonAndFieldBar() {
+  }
+
+  private void onRegisterDialogConfirmed(ConfirmEvent confirmEvent) {
+    Data data = confirmEvent.getData();
```
I think the readability could be improved by not getting "data" but something more explicit like "BatchRegistrationContent"
```suggestion
    Data data = confirmEvent.getData();
```

Steffengreiner commented on user-interface/src/main/java/life/qbic/datamanager/views/projects/project/samples/SampleDetailsComponent.java:
```diff
@@ -282,9 +348,49 @@ private void addEventListeners() {
+              batchRegistrationContent.isPilot(), context.projectId().orElseThrow())
+          .onError(responseCode -> displayRegistrationFailure())
+          .onValue(batchId -> displayRegistrationSuccess());
+    }
```
Either use these methods in the onRegisterDialogConfirmed method or remove them 🗑️

KochTobi replied:
They are now removed :)

Steffengreiner commented on user-interface/src/main/java/life/qbic/datamanager/views/projects/project/samples/registration/batch/BatchRegistrationDialog.java:
```diff
@@ -241,4 +235,15 @@ public BatchRegistrationContent batchRegistrationContent() {
+    public CancelEvent(BatchRegistrationDialog source, boolean fromClient) {
+      super(source, fromClient);
+    }
+  }
```
Is this distinction really necessary? Where is the cancelevent employed?

KochTobi replied:
I don't know. The whole dialog is removed now.

Steffengreiner commented on user-interface/src/main/java/life/qbic/datamanager/views/projects/project/samples/registration/batch/BatchRegistrationDialog2.java:
```diff
@@ -0,0 +1,547 @@
+ * <p>detailed description</p>
+ *
+ * @since <version tag>
+ */
```
Missing JD

Steffengreiner commented on user-interface/src/main/java/life/qbic/datamanager/views/projects/project/samples/registration/batch/BatchRegistrationDialog2.java:
```diff
@@ -0,0 +1,547 @@
+ *
+ * @since <version tag>
+ */
+public class BatchRegistrationDialog2 extends DialogWindow {
```
Adapt naming schema after removing the old dialog
```suggestion
public class BatchRegistrationDialog extends DialogWindow {
```

Steffengreiner commented on user-interface/src/main/java/life/qbic/datamanager/views/projects/project/samples/registration/batch/BatchRegistrationDialog2.java:
```diff
@@ -0,0 +1,547 @@
+  private final List<Analyte> analytes;
+
+
+  public BatchRegistrationDialog2(String experimentName,
```
NIT: The constructor does a lot of work here, maybe the individual initialization could be split up into distinct methods(initSpreadsheet, initSpreadsheetControls etc.) so it's more readable 🤔

KochTobi replied:
nope makes it harder to read. This can all be in the same scope IMHO.

Steffengreiner commented on user-interface/src/main/java/life/qbic/datamanager/views/projects/project/samples/registration/batch/BatchRegistrationDialog2.java:
```diff
@@ -0,0 +1,547 @@
+  }
+
+  private void onAddRowClicked(ClickEvent<Button> clickEvent) {
+    spreadsheet.setValidationMode(ValidationMode.LAZY);
```
I suspect this is a trick so the validation does not show errors if the row is added? 
```suggestion
    spreadsheet.setValidationMode(ValidationMode.LAZY);
```

KochTobi replied:
The validation is turned off, and on again after the new row is present 😄

Steffengreiner commented on user-interface/src/main/java/life/qbic/datamanager/views/projects/project/samples/registration/batch/BatchRegistrationDialog2.java:
```diff
@@ -0,0 +1,547 @@
+    }
+    if (batchNameField.isInvalid()) {
+      return;
+    }
```
Should there be a notification shown to the user informing him that there are invalid fields?

KochTobi replied:
As the fields are highlighted and I did not get this input, this is up to the user tests.

Steffengreiner commented on user-interface/src/main/java/life/qbic/datamanager/views/projects/project/samples/registration/batch/BatchRegistrationDialog2.java:
```diff
@@ -0,0 +1,547 @@
+    public CancelEvent(BatchRegistrationDialog2 source, boolean fromClient) {
+      super(source, fromClient);
+    }
+  }
```
Is a seperate CancelEvent necessary or can the vaadin event be used?

KochTobi replied:
Are you referring to the ConfirmDialog Cancel Event? As this is not a confirm dialog, there is no appropriate event here.

Steffengreiner commented on user-interface/src/main/java/life/qbic/datamanager/views/projects/project/samples/registration/batch/BatchRegistrationDialog2.java:
```diff
@@ -0,0 +1,547 @@
+    }
+  }
+
+  public static class SampleInfo {
```
Major NIT: 
```suggestion
  public static class SampleEntry {
```

KochTobi replied:
As this is the information regarding a sample and the sample is not yet entered anywhere, I would prefer to keep SampleInfo

Steffengreiner commented on user-interface/src/main/java/life/qbic/datamanager/views/projects/project/samples/registration/batch/BatchRegistrationDialog2.java:
```diff
@@ -0,0 +1,547 @@
+
+    public Data getData() {
+      return data;
+    }
```
NIT: I'd recommend to make this naming scheme more explicit

Steffengreiner reviewed (changes requested):
Heya @KochTobi, Thanks for the awesome contribution! I found some minor inconsistencies in the code which could be addressed. 
However the bigger concern is the styling of the dialog and the surrounding page which should be addressed

1.) Duplicate register button below sample grid
![Bildschirmfoto 2023-11-20 um 12 17 54](https://github.com/qbicsoftware/data-manager-app/assets/29627977/de58bedd-64bc-4c49-8afa-30e75cb1e3d0)

2.) Missing spacing between the components within the dialog. 
![Bildschirmfoto 2023-11-20 um 12 18 23](https://github.com/qbicsoftware/data-manager-app/assets/29627977/119f6655-37a3-4057-8f10-d7b5b8727442)

3.) Column header overflowing into other columns and spreadsheet columns don't take full width
![Bildschirmfoto 2023-11-20 um 12 18 25](https://github.com/qbicsoftware/data-manager-app/assets/29627977/f1d90cda-56c6-43b5-96b2-03e837afc53f)

4.) Cell content overflowing into next column 
![Bildschirmfoto 2023-11-20 um 12 18 33](https://github.com/qbicsoftware/data-manager-app/assets/29627977/f21eb83c-da75-4a41-899b-54c76e0b2875)

5.) Funnily enough if a species, analyte or specimen ending with a number is dragged down the number is incremented which leads to an immediate validation error
![Bildschirmfoto 2023-11-20 um 12 18 48](https://github.com/qbicsoftware/data-manager-app/assets/29627977/e92ef36a-e300-49c8-9de9-c5dc773253a4)

6.) Replicate is not selectable until a condition has been selected 
![Bildschirmfoto 2023-11-20 um 12 19 25](https://github.com/qbicsoftware/data-manager-app/assets/29627977/3922c28f-3d90-40b0-ba3a-8ebd000c2c80)

7.) Finally the dialog look does not match the Figma Prototype concerning button styles in the footer, seperation lines between the dialogs, header style and text content. 
![Bildschirmfoto 2023-11-20 um 12 22 29](https://github.com/qbicsoftware/data-manager-app/assets/29627977/1d516668-7f51-4e4d-92bd-94d3fa05ee1b)

KochTobi commented on user-interface/frontend/themes/datamanager/components/dialog.css:
```diff
@@ -151,14 +156,29 @@ vaadin-dialog-overlay::part(title) {
+}
+
+.batch-registration-dialog .batch-controls > .batch-name-field {
+  width: calc(100% / 3);
```
nope that is not the same. 1/3 are 33.33333333333... %

KochTobi commented:
Dear @Steffengreiner 
Thank you for your review!

> 1.) Duplicate register button below sample grid !

This is resolved. The button was added and replaced as I removed the old code. 👍

> 2.) Missing spacing between the components within the dialog.

The spacing is not final yet as this takes a lot of time compared to the benefit. This can be tackled separately.

> 3.) Column header overflowing into other columns and spreadsheet columns don't take full width

Without content, scaling of the columns is difficult to do. Columns adjust their width to what is needed when filled with content.

> 4.) Cell content overflowing into next column ![Bildschirmfoto 2023-11-20 um 12 18 33](https://user-images.githubusercontent.com/29627977/284249352-f21eb83c-da75-4a41-899b-54c76e0b2875.png)

Can you provide me with the view port width you were using? I cannot reproduce this behaviour.

> 5.) Funnily enough if a species, analyte or specimen ending with a number is dragged down the number is incremented which leads to an immediate validation error

This is a `feature` of the vaadin spreadsheet. Nothing we can do about it, sadly.

> 6.) Replicate is not selectable until a condition has been selected

Yes, as the replicates are filled from the group which is selected by the condition, I do not know how to fix this.

> 7.) Finally the dialog look does not match the Figma Prototype concerning button styles in the footer, seperation lines between the dialogs, header style and text content.

I did not touch the DialogWindow class. The current implementation of the RegisterDialog matches the other dialogs in the application. I think the dialog styling should be tackled separately by using the DialogWindow class or modifying the CSS.

Steffengreiner commented:
Heya @KochTobi 

Thanks for the quick update, however i found an additional concern:  

1.) The dialog should not be openable for experiments without defined groups and variables
![Bildschirmfoto 2023-11-20 um 14 58 20](https://github.com/qbicsoftware/data-manager-app/assets/29627977/a9dfa84e-cad4-4a48-8bd3-d9f06c2c11a0)
![Bildschirmfoto 2023-11-20 um 14 58 32](https://github.com/qbicsoftware/data-manager-app/assets/29627977/4089ab9c-48cc-4609-9b33-6e69467f98da)

Addressing your questions: 

**A) Overflowing content.** 
My viewport size is 1920 width and 1200 height(default monitor size in our office) 
![Bildschirmfoto 2023-11-20 um 14 46 23](https://github.com/qbicsoftware/data-manager-app/assets/29627977/4d9b0f98-d323-4f61-95b3-ee143ce03afb)
This needs to be addressed even if we have to set default min-widths. 
The title of a column should not overflow into the next column. 
Our previous solution determined the longest string in the selectable content and adjusted the width accordingly. 

**B) Bio replicates and condition. **
From a UI perspective this is a major issue since the user selects the replicate before the condition, which is not possible anymore. I'd recommend to either: 

1.) Escalate the issue with SF so the replicates are predetermined in future. 
2.) Swap the columns so the user selects the condition first 
3.) Inform the user that he has to select the condition first before she's able to select the replicate. (Maybe via a custom comment?

**C) Auto incrementation of numbers.** 
This should also be addressed but the priority is lower, we previously had a hacky fix by adding an empty space at the end of these cells, which can be trimmed easily

Steffengreiner reviewed (approved):
Nicely done. 👍
I still have some suggestions on how the dialog could be improved but they can be tackled in a seperate PR
